File: sdk/python/tap/x402/payment.py
"""x402 `X-PAYMENT` payload for opening a TAP channel.

The payload carries the parameters the server needs to know about the
session even before the channel-open transaction is observed on-chain
(session key, deposit, agreed prices, prepaid input floor, nonce). The
actual transaction is base64-encoded so x402 facilitators can submit it
verbatim."""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Any

from tap.exceptions import X402Error


@dataclass(frozen=True, slots=True)
class OpenChannelPayment:
    """Consumer-side x402 payment payload for `tap.v1.channel`."""

    scheme: str
    network: str
    consumer_pubkey: str
    session_key: str
    nonce: int
    deposit_micro: int
    input_price_micro: int
    output_price_micro: int
    prepaid_input_micro: int
    duration_secs: int
    dispute_secs: int
    trailing_buffer_tokens: int
    transaction_b64: str

    def to_payload(self) -> dict[str, Any]:
        return {
            "scheme": self.scheme,
            "network": self.network,
            "extra": {
                "consumer_pubkey": self.consumer_pubkey,
                "session_key": self.session_key,
                "nonce": self.nonce,
                "deposit_micro": self.deposit_micro,
                "input_price_micro": self.input_price_micro,
                "output_price_micro": self.output_price_micro,
                "prepaid_input_micro": self.prepaid_input_micro,
                "duration_secs": self.duration_secs,
                "dispute_secs": self.dispute_secs,
                "trailing_buffer_tokens": self.trailing_buffer_tokens,
                "transaction": self.transaction_b64,
            },
        }


def encode_payment(payment: OpenChannelPayment) -> str:
    raw = json.dumps(payment.to_payload(), separators=(",", ":")).encode("utf-8")
    return base64.b64encode(raw).decode("ascii")


def decode_payment(header_value: str) -> OpenChannelPayment:
    try:
        payload = json.loads(base64.b64decode(header_value, validate=True))
    except (ValueError, json.JSONDecodeError) as exc:
        raise X402Error("X-PAYMENT is not valid base64-JSON") from exc

    extra = payload.get("extra") or {}
    try:
        return OpenChannelPayment(
            scheme=payload["scheme"],
            network=payload["network"],
            consumer_pubkey=extra["consumer_pubkey"],
            session_key=extra["session_key"],
            nonce=int(extra["nonce"]),
            deposit_micro=int(extra["deposit_micro"]),
            input_price_micro=int(extra["input_price_micro"]),
            output_price_micro=int(extra["output_price_micro"]),
            prepaid_input_micro=int(extra["prepaid_input_micro"]),
            duration_secs=int(extra["duration_secs"]),
            dispute_secs=int(extra["dispute_secs"]),
            trailing_buffer_tokens=int(extra["trailing_buffer_tokens"]),
            transaction_b64=str(extra["transaction"]),
        )
    except (KeyError, ValueError, TypeError) as exc:
        raise X402Error("X-PAYMENT payload is missing or malformed fields") from exc

File: sdk/typescript/src/x402/payment.ts
/**
 * x402 `X-PAYMENT` payload for opening a TAP channel. Wire format mirrors
 * `tap.x402.payment` in the Python SDK: outer `{scheme, network, extra}`
 * with all TAP-specific fields under `extra`, base64-encoded JSON in the
 * HTTP header.
 */

import { X402Error } from "../exceptions.js";
import { _internal } from "../protocol/codec.js";

export interface OpenChannelPayment {
  readonly scheme: string;
  readonly network: string;
  readonly consumerPubkey: string;
  readonly sessionKey: string;
  readonly nonce: bigint;
  readonly depositMicro: bigint;
  readonly inputPriceMicro: bigint;
  readonly outputPriceMicro: bigint;
  readonly prepaidInputMicro: bigint;
  readonly durationSecs: number;
  readonly disputeSecs: number;
  readonly trailingBufferTokens: number;
  /** Fully-signed channel-open transaction, base64 of the serialized bytes. */
  readonly transactionB64: string;
}

interface ExtraPayload {
  consumer_pubkey: string;
  session_key: string;
  nonce: number;
  deposit_micro: number;
  input_price_micro: number;
  output_price_micro: number;
  prepaid_input_micro: number;
  duration_secs: number;
  dispute_secs: number;
  trailing_buffer_tokens: number;
  transaction: string;
}

interface RootPayload {
  scheme: string;
  network: string;
  extra: ExtraPayload;
}

function bigintToNumber(value: bigint, field: string): number {
  if (value > BigInt(Number.MAX_SAFE_INTEGER) || value < 0n) {
    throw new RangeError(`x402 payment field ${field} (${value}) exceeds Number.MAX_SAFE_INTEGER`);
  }
  return Number(value);
}

function toPayload(p: OpenChannelPayment): RootPayload {
  return {
    scheme: p.scheme,
    network: p.network,
    extra: {
      consumer_pubkey: p.consumerPubkey,
      session_key: p.sessionKey,
      nonce: bigintToNumber(p.nonce, "nonce"),
      deposit_micro: bigintToNumber(p.depositMicro, "deposit_micro"),
      input_price_micro: bigintToNumber(p.inputPriceMicro, "input_price_micro"),
      output_price_micro: bigintToNumber(p.outputPriceMicro, "output_price_micro"),
      prepaid_input_micro: bigintToNumber(p.prepaidInputMicro, "prepaid_input_micro"),
      duration_secs: p.durationSecs,
      dispute_secs: p.disputeSecs,
      trailing_buffer_tokens: p.trailingBufferTokens,
      transaction: p.transactionB64,
    },
  };
}

export function encodePayment(payment: OpenChannelPayment): string {
  return _internal.utf8ToBase64(JSON.stringify(toPayload(payment)));
}

export function decodePayment(headerValue: string): OpenChannelPayment {
  let payload: RootPayload;
  try {
    payload = JSON.parse(_internal.base64ToUtf8(headerValue)) as RootPayload;
  } catch (cause) {
    throw new X402Error("X-PAYMENT is not valid base64-JSON", { cause: cause as Error });
  }
  const extra = payload.extra;
  if (!extra) {
    throw new X402Error("X-PAYMENT payload is missing or malformed fields");
  }
  try {
    return {
      scheme: payload.scheme,
      network: payload.network,
      consumerPubkey: extra.consumer_pubkey,
      sessionKey: extra.session_key,
      nonce: BigInt(extra.nonce),
      depositMicro: BigInt(extra.deposit_micro),
      inputPriceMicro: BigInt(extra.input_price_micro),
      outputPriceMicro: BigInt(extra.output_price_micro),
      prepaidInputMicro: BigInt(extra.prepaid_input_micro),
      durationSecs: Number(extra.duration_secs),
      disputeSecs: Number(extra.dispute_secs),
      trailingBufferTokens: Number(extra.trailing_buffer_tokens),
      transactionB64: String(extra.transaction),
    };
  } catch (cause) {
    throw new X402Error("X-PAYMENT payload is missing or malformed fields", {
      cause: cause as Error,
    });
  }
}
